core/src/main/java/org/bitcoinj/evolution/AbstractQuorumState.java

```diff
@@ -0,0 +1,604 @@
+import org.bitcoinj.core.MasternodeSync;
+import org.bitcoinj.core.Message;
+import org.bitcoinj.core.NetworkParameters;
+import org.bitcoinj.core.Peer;
+import org.bitcoinj.core.PeerGroup;
+import org.bitcoinj.core.Sha256Hash;
+import org.bitcoinj.core.StoredBlock;
+import org.bitcoinj.core.Utils;
+import org.bitcoinj.core.VerificationException;
+import org.bitcoinj.core.listeners.ChainDownloadStartedEventListener;
+import org.bitcoinj.core.listeners.NewBestBlockListener;
+import org.bitcoinj.core.listeners.PeerConnectedEventListener;
+import org.bitcoinj.core.listeners.PeerDisconnectedEventListener;
+import org.bitcoinj.core.listeners.ReorganizeListener;
+import org.bitcoinj.quorums.LLMQParameters;
+import org.bitcoinj.quorums.SimplifiedQuorumList;
+import org.bitcoinj.store.BlockStore;
+import org.bitcoinj.store.BlockStoreException;
+import org.bitcoinj.utils.Threading;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.FileNotFoundException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * The abstract base class of all Quorum State classes.  This class handles sync of masternode and quorum lists
+ *
+ * @param <Request>     A class derived from {@link AbstractQuorumRequest} used as the message to request updates
+ *                      to the masternode and quorum lists
+ * @param <DiffMessage> A class derived from {@link AbstractDiffMessage} used as the message that contains updates
+ *                      *           to the masternode and quorum lists
+ */
+
+public abstract class AbstractQuorumState<Request extends AbstractQuorumRequest, DiffMessage extends AbstractDiffMessage> extends Message {
+
+    public static final int SNAPSHOT_LIST_PERIOD = 576; // once per day
+    public static final int LISTS_CACHE_SIZE = 576;
+    public static final int SNAPSHOT_TIME_PERIOD = 60 * 60 * 26;
+
+    public static int MAX_CACHE_SIZE = 10;
+    public static int MIN_CACHE_SIZE = 1;
+
+    private static final Logger log = LoggerFactory.getLogger(AbstractQuorumState.class);
+    protected ReentrantLock lock = Threading.lock("AbstractQuorumState");
+
+    Context context;
+    AbstractBlockChain headerChain;
+    AbstractBlockChain blockChain;
+    BlockStore headerStore;
+    BlockStore blockStore;
+
+    QuorumUpdateRequest<Request> lastRequest;
+
+    static final long WAIT_GETMNLISTDIFF = 5000;
+    Peer downloadPeer;
+    boolean waitingForMNListDiff;
+    boolean initChainTipSyncComplete = false;
+    LinkedHashMap<Sha256Hash, StoredBlock> pendingBlocksMap;
+    ArrayList<StoredBlock> pendingBlocks;
+
+    int failedAttempts;
+    static final int MAX_ATTEMPTS = 10;
+
+    public SimplifiedMasternodeListManager.SyncOptions syncOptions;
+    public int syncInterval;
+
+    QuorumStateManager stateManager;
```

Comment:
Most of these class members were taken out of `SimplifiedMasternodeListManager` as part of the refactoring to create a base class that handles most of the similarities between DIP4 and DIP24 quorum list management.